### ISSUE TITLE
VideoStandard needed in video table

### ISSUE THREAD

**Issue by IvonArroyo** [center-for-knowledge-communication/mathspring#313]
Boming and I talked about him allowing to upload the video links, and create new videos into the video table.

This led to a whole discussion, as currently videos are assigned to problems, and in some cases there are two videos for problems corresponding to the same standard.

Authors should be able to assign a video to a problem (from the problem authoring tab) and they will be able to "see" the videos assigned to a standard, from the standards tab.

We came up with a potential solution, but the solution involves adding a **videostandard** field in the video table. When we upload  the link to a video, we have to say what standard that video corresponds to.

So we need 2 different things here:

1) Can somebody add a new column called videostandard (or alike) to the video table, something that defaults to null

2) Could Tom fill in the videostandard column for all the videos that already exist in the video table? This can be done by looking a the problem table. The videoid should be assigned to problems, which have a standard associated to it.

3) After all of this is done, we might want to change the video selector in the following way, but only after all of the previous is done: "when the video selector looks for a video for the current problem, it first looks if there is a videoid assigned to the problem. If it is null, it should look for a video in the video table that has the standard that corresponds to this current problem".

--Ivon

**Comment by perspegrity5** [comment]
To make sure I understand, is this correct Ivon?:

- A standard has one (primary) video.  
- If a problem does not have a video associated with it, it will use the standard's video.

-OR- do you mean the following:
- A standard is associated with a SET of videos
- If a problem does not have a video associated with it, it will choose one of the standard's videos (randomly choose one that has not been seen yet).

**Comment by perspegrity5** [comment]
New Status:
- new table was created: video_standard_map
- video_standard_map was populated with 189 rows based on problems that pointed to videos; resulting in 84 standards having one or more videos
- I noted in the sysNotes field of the map table that these originated in problems.
- I am attaching an excel file listing all video objects, showing which do not have any links to problems (thus, or standards). Look for "#NA" for orphaned videos. NOTE: this is only done for Ready/Testable/In-Progress problems, so some videos may be linked to dead problems that won't show up here.
- (Note that I included fields segmentStartSeconds segmentEndSeconds, in case we find videos where only a segment of it applies to the specified standard. This is for future use, ignore for now.)

NEXT STEPS: MS Coding:
- alter video selection routine: if a problem points to a video use that; if not, use the problem's standard and look at the mapping created above.  Choose randomly from that standard's videos.  
- (Future implementation: if there is more than one vid for the standard, show a popup of the Video's name.  But not all videos names have been entered, so we have to do that first)

Next Steps: Authoring Tool coding:
- We need to integrate this into both the problem authoring and standards authoring.
- Problem authoring: allow for (1) show existing video; (2) add: choose from existing videos (list them by name); (3) add: enter a new video ID.
- Standard authoring: allow for (1) show existing video (choose from list from map); (2) add: choose from existing videos (list them by name); (3) add: enter a new video ID 
- For both of the above "#3 create new video" -- this should first check if the entered URL is already existing and flag that. It should also allow the author to enter the name and description (and URL) of the video.

Next steps: Content authors:
- All used videos should have a name and short description.
- Decide if the un-used videos belong to some standard.

**Comment by FrankSylvia** [comment]
Testing instructions as of 7/12/2021

Testable:
alter video selection routine: if a problem points to a video use that; if not, use the problem's standard and look at the mapping created above. 

TBD:
There are currently no standards with multiple videos.  The following features are on hold until the some standards have multiple videos.

Choose randomly from that standard's videos.
(Future implementation: if there is more than one vid for the standard, show a popup of the Video's name. But not all videos names have been entered, so we have to do that first)
